### PR TITLE
feature: implemented ngx.req module and its function add_header().

### PR DESCRIPTION
**Pull request by spacewander**
The `ngx.req.add_header` works mostly like `ngx.req.set_header`.
However, unlike `ngx.req.set_header`, this method appends new header to the old
one instead of overriding it if the header is not builtin.

For the builtin header, this method overrides it instead of appending to it.
IMHO, it is reasonable since appending to the builtin header like `User-Agent`
is confusing and semantically wrong.

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

**Comment by spacewander** [comment]
This is the sister pr of https://github.com/openresty/lua-resty-core/pull/213

File: src/ngx_http_lua_headers.c [hunk 1]
```diff
@@ -1322,9 +1322,13 @@ ngx_http_lua_ffi_set_resp_header(ngx_http_request_t *r, const u_char *key_data,
 
 
 int
-ngx_http_lua_ffi_req_header_set_single_value(ngx_http_request_t *r,
-    const u_char *key, size_t key_len, const u_char *value, size_t value_len)
+ngx_http_lua_ffi_req_set_header(ngx_http_request_t *r,
```
Comment (thibaultcha) [inline src/ngx_http_lua_headers.c]: We should align this function's name with `ngx_http_lua_ffi_set_resp_header`. It should thus be `ngx_http_lua_ffi_set_req_header`.

File: src/ngx_http_lua_headers.c [hunk 2]
```diff
@@ -1345,22 +1349,51 @@ ngx_http_lua_ffi_req_header_set_single_value(ngx_http_request_t *r,
...
+                }
+
+                ngx_memcpy(v.data, p, len);
+                v.data[len] = '\0';
+                v.len = len;
+
+                if (ngx_http_lua_set_input_header(r, k, v, override)
+                    != NGX_OK)
+                {
+                    return NGX_ERROR;
```
Comment (thibaultcha) [inline src/ngx_http_lua_headers.c]: `ngx_http_lua_ffi_set_resp_header` has a slightly different implementation. I believe this implementation should closely mirror it. Additionally, `ngx_http_lua_ffi_set_resp_header` also sets error messages on "failed to set header" or "no memory" errors. I think that we should do the same here.

**Comment by thibaultcha** [comment]
Merged!

**Pull request closed without merging**